presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java

```diff
@@ -561,6 +545,75 @@ private void invokeNoMoreSplitsIfNecessary()
         return COMPLETED_FUTURE;
     }
 
+    private Iterator<InternalHiveSplit> buildManifestFileIterator(Table table,
+            FileStatus symlinkFile,
+            FileSystem fileSystem,
+            DirectoryLister directoryLister,
+            NamenodeStats namenodeStats,
+            InternalHiveSplitFactory splitFactory,
+            boolean splittable)
+    {
+        // Strategy: List once, filter file listing based on manifest file. Complication is that not all files in the manifest are guaranteed
+        // to have a common root.
+
+        List<Iterator<LocatedFileStatus>> fileStatusIterators = new ArrayList<>();
+        List<Path> paths = new ArrayList<>();
+
+        try (BufferedReader reader = new BufferedReader(new InputStreamReader(fileSystem.open(symlinkFile.getPath()), StandardCharsets.UTF_8))) {
+            CharStreams.readLines(reader).stream()
+                    .map(Path::new)
+                    .forEach(paths::add);
+        }
+        catch (IOException e) {
+            throw new PrestoException(HIVE_BAD_DATA, "Error parsing symlinks from: " + symlinkFile, e);
+        }
+
+        ImmutableSet<Path> parents = paths.stream()
+                .map(Path::getParent)
+                .distinct()
+                .collect(toImmutableSet());
+
+        for (Path parent : parents) {
```

Comment:
could it be that some parent is child of another parent (e.g parents are nested)? Would splits be enumerated twice then?

Reply:
It is possible, but only if recursive directory listing is enabled. I added a check to make sure it is not.

Reply:
This now falls back to the original codepath if recursive listing is enabled or if there are multiple parents.